pallets/stablewap/src/lib.rs

```diff
@@ -0,0 +1,662 @@
+		InsufficientLiquidity,
+
+		/// Insufficient liquidity left in the pool after withdrawal.
+		InsufficientLiquidityRemaining,
+
+		/// Amount is less than min trading limit.
+		InsufficientTradingAmount,
+
+		/// Minimum limit has not been reached during trade.
+		BuyLimitNotReached,
+
+		/// Maximum limit has been exceeded during trade.
+		SellLimitExceeded,
+
+		/// Initial liquidity of asset must be > 0.
+		InvalidInitialLiquidity,
+
+		/// Account balance is too low.
+		BalanceTooLow,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Create a 2-asset pool with initial liquidity for both assets.
+		///
+		/// Both assets must be correctly registered in `T::AssetRegistry`
+		///
+		/// Initial liquidity must be > 0.
+		///
+		/// Origin is given corresponding amount of shares.
+		///
+		/// Parameters:
+		/// - `origin`: Must be T::CreatePoolOrigin
+		/// - `assets`: Asset ids tuple
+		/// - `initial_liquidity`: Corresponding initial liquidity of `assets`
+		/// - `amplification`: Pool amplification
+		/// - `fee`: trade fee to be applied in sell/buy trades
+		///
+		/// Emits `PoolCreated` event when successful.
+		#[pallet::weight(<T as Config>::WeightInfo::create_pool())]
+		#[transactional]
+		pub fn create_pool(
+			origin: OriginFor<T>,
+			assets: (T::AssetId, T::AssetId),
+			initial_liquidity: (Balance, Balance),
+			amplification: Balance,
+			fee: Permill,
+		) -> DispatchResult {
+			let who = T::CreatePoolOrigin::ensure_origin(origin)?;
+
+			let pool_assets: PoolAssets<T::AssetId> = assets.into();
+
+			ensure!(pool_assets.is_valid(), Error::<T>::SameAssets);
+
+			for asset in (&pool_assets).into_iter() {
+				ensure!(T::AssetRegistry::exists(asset), Error::<T>::AssetNotRegistered);
+			}
+
+			ensure!(
+				initial_liquidity.0 > Balance::zero(),
+				Error::<T>::InvalidInitialLiquidity
+			);
+			ensure!(
+				initial_liquidity.1 > Balance::zero(),
+				Error::<T>::InvalidInitialLiquidity
+			);
+
+			ensure!(
+				T::Currency::free_balance(assets.0, &who) >= initial_liquidity.0,
+				Error::<T>::BalanceTooLow,
+			);
+			ensure!(
+				T::Currency::free_balance(assets.1, &who) >= initial_liquidity.1,
+				Error::<T>::BalanceTooLow,
+			);
+
+			let share_asset_ident = T::ShareAccountId::name(&pool_assets, Some(POOL_IDENTIFIER));
+
+			let share_asset = T::AssetRegistry::get_or_create_shared_asset(
+				share_asset_ident,
+				(&pool_assets).into(),
+				Balance::zero(),
+			)?;
+
+			let pool_id = PoolId(share_asset);
+
+			let pool = Pools::<T>::try_mutate(
+				&pool_id,
+				|maybe_pool| -> Result<PoolInfo<T::AssetId, Balance>, DispatchError> {
+					ensure!(maybe_pool.is_none(), Error::<T>::PoolExists);
+
+					let pool = PoolInfo {
+						assets: pool_assets.clone(),
+						amplification,
+						fee,
+					};
+
+					*maybe_pool = Some(pool.clone());
+
+					Ok(pool)
+				},
+			)?;
+
+			let reserves: AssetAmounts<Balance> = initial_liquidity.into();
+
+			let share_amount = calculate_add_liquidity_shares(
```

Comment:
does this fail if wrong amplification is setup? Don't see any ensures (but see it's transactional)

Reply:
if math fails - it fails of course. 

what could we check ?

Reply:
If the amplification is in the bounds that you have set in the config.

Reply:
Oh sorry if the amplification is in the bounds that we allow for and is benchmarked for.

Reply:
We don't have currently min/max amp bounds. Do you think it is needed as pallet's configurable property?

Reply:
I would tend to think so since it can change the results of the math and we test for it. I would like to see opinion from @poliwop. It might not be a big issue when there is committee to assess proper config. I would expect that you can't set some values (or at least see tests the extreme values won't break it and include in benchmarks / fuzz)

Reply:
I would recommend setting a min of A = 2, max of A = 10,000, and testing at these boundaries.